suites/tipset/rewards_penalties.go

```diff
@@ -208,6 +209,64 @@ func TestMinerRewardsAndPenalties(t *testing.T, factory state.Factories) {
 		td.AssertBalance(builtin.BurntFundsActorAddr, big.Add(halfBalance, gasPenalty))
 	})
 
+	t.Run("insufficient gas to cover return value", func(t *testing.T) {
+		td := builder.Build(t)
+		defer td.Complete()
+
+		miner := td.ExeCtx.Miner
+		tb := drivers.NewTipSetMessageBuilder(td)
+
+		alice, _ := td.NewAccountActor(drivers.BLS, acctDefaultBalance)
+
+		retval := chain.MustSerialize(&miner_spec.GetControlAddressesReturn{
+			Owner:  td.StateDriver.BuiltinMinerInfo().OwnerID,
+			Worker: td.StateDriver.BuiltinMinerInfo().WorkerID,
+		})
+
+		prevRewards1 := td.GetRewardSummary()
+
+		// get a successful result so we can determine how much gas it costs. We'll reduce this by 1 in a subsequent call
+		// to test insufficient gas to cover return value.
+		result1 := tb.WithBlockBuilder(
+			drivers.NewBlockBuilder(td, td.ExeCtx.Miner).
+				WithBLSMessageAndRet(td.MessageProducer.MinerControlAddresses(miner, alice, nil, chain.Nonce(0)),
+					retval, // required to satisfy testing methods, unrelated to current test.
+				),
+		).ApplyAndValidate()
+		newRewards1 := td.GetRewardSummary()
+		validateRewards(td, prevRewards1, newRewards1, miner, result1.Receipts[0].GasUsed.Big(), big.Zero())
+
+		tb.Clear()
+		/* now the test */
+
+		prevRewards2 := td.GetRewardSummary()
+		td.ExeCtx.Epoch++
+
+		// get the gas used
+		gasUnitsUsed := result1.Receipts[0].GasUsed
+		gasLimit := int64(gasUnitsUsed) - 1
+		// apply the message again with a reduced gasLimit
+		result2 := tb.WithBlockBuilder(
+			drivers.NewBlockBuilder(td, td.ExeCtx.Miner).
+				WithBLSMessageAndCode(td.MessageProducer.MinerControlAddresses(miner, alice,
+					nil, chain.Nonce(1), chain.GasPrice(1), chain.GasLimit(gasLimit)),
+					exitcode.SysErrOutOfGas,
+				),
+		).ApplyAndValidate()
+		newRewards2 := td.GetRewardSummary()
+
+		// ensure the actual gas charged is equal to the gas limit rather than the amount consumed up to but excluding
+		// the return value which is smaller than the gas limit.
+		assert.Equal(t, gasLimit, int64(result2.Receipts[0].GasUsed))
+
+		thisRwd := big.Add(newRewards2.LastPerEpochReward, result2.Receipts[0].GasUsed.Big())
+		prevRwd := big.Add(prevRewards2.LastPerEpochReward, result1.Receipts[0].GasUsed.Big())
+
+		assert.Equal(td.T, big.Add(prevRwd, thisRwd), td.GetBalance(miner))
+		assert.Equal(td.T, big.Sub(prevRewards2.Treasury, thisRwd), newRewards2.Treasury)
```

Comment:
This isn't quite right. The gas cost isn't subtracted from the reward actor, only the block reward.